### PR TITLE
1766 - Git Tag and Chart Version assumed to match

### LINKED ISSUE #1766: Git Tag and Chart Version assumed to match
### Steps to reproduce
Create a zarf.yaml with this component:

```yaml
  - name: dcgm-exporter
    required: true
    charts:
    - name: dcgm-exporter
      version: 3.1.7-3.1.4 # <- this tag doesn't match the chart version
      url: https://github.com/NVIDIA/dcgm-exporter.git
      gitPath: deployment
      namespace: "leapfrogai"
    images:
      - "nvcr.io/nvidia/k8s/dcgm-exporter:3.1.7-3.1.4-ubuntu20.04"
   ```

This particular application has a mismatch between the git tag and the chart version.  This causes an error in either one of two ways since those two are assumed to match:

1. Build error: If using the version matching the chart version, the package doesn't build since Zarf can't find the Git Tag
2. Deploy Error:  If using the git tag, zarf can't deploy the chart since it can't find a chart with the provided version

### Expected result

Be able to deploy an arbitrary chart version from a git repository

### Additional Context
Add any other context or screenshots about the technical debt here.

Ideas/for implementation:

1. Add new property to the `charts` that allows for setting `gitVersion` separately from `version`, but have the default `gitVersion` match the `version` field
2. Add more robust regex on the `url` to allow for passing branch/tag versions in the URL.  Some ambiguity here on backwards compatibility.  If nothing is set, does it default to the default branch, or the `version`?
3. Don't have zarf check to ensure the version matches when deploying a chart from git.

### PR DESCRIPTION
**Pull request by cmwylie19**
## Description

What we could do (to keep backwards compat in tact and address https://github.com/defenseunicorns/zarf/issues/1766) is see if there is an @ ref on the git URL and if there is, leave it on there, and if not append the @tag as we do today. Dedup some code in src/pkg/packager/prepare.go, and src/pkg/packager/create.go.

## Related Issue

Fixes #1766

Relates to 

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Other (security config, docs update, etc)

## Checklist before merging

- [x] Test, docs, adr added or updated as needed
- [x] [Contributor Guide Steps](https://github.com/defenseunicorns/zarf/blob/main/CONTRIBUTING.md#developer-workflow) followed

### CODE REVIEW

**Comment by cmwylie19** [comment]
I am removing the `needs-tests` label because I am going to address the tests in the [metrics pr](https://github.com/defenseunicorns/zarf/pull/1853) as this current PR is a prereq along with the [named port pr](https://github.com/defenseunicorns/zarf/pull/1894) and the waitFor pr which has been merged.

File: src/internal/packager/helm/repo.go [hunk 1]
```diff
@@ -22,65 +24,81 @@ import (
 	"helm.sh/helm/v3/pkg/repo"
 )
 
+func (h *Helm) PackageChart(destination string) error {
```
Comment (Racer159) [inline src/internal/packager/helm/repo.go:27]: lint

File: src/internal/packager/helm/repo.go [hunk 2]
```diff
@@ -22,65 +24,81 @@ import (
...
+	if len(h.Chart.URL) > 0 {
+		url, refPlain, err := transform.GitTransformURLSplitRef(h.Chart.URL)
+		// check if the chart is a git url with a ref (if an error is returned url will be empty)
+		isGitURL := strings.HasSuffix(url, ".git")
+		if err != nil {
+			message.Debugf("unable to parse the url, continuing with %s", h.Chart.URL)
+		}
+
+		if isGitURL {
+			// if it is git url append chart version as if its a tag
```
Comment (Racer159) [inline src/internal/packager/helm/repo.go:37]:
```suggestion
			// if it is a git url append chart version as if its a tag
```

**Review by Racer159 — CHANGES_REQUESTED** [review]

File: src/internal/packager/helm/repo.go (hunk 1 above)
Comment (cmwylie19) [reply src/internal/packager/helm/repo.go:28]: thanks, done!

File: src/internal/packager/helm/repo.go [hunk 3]
```diff
@@ -22,65 +24,82 @@ import (
 	"helm.sh/helm/v3/pkg/repo"
 )
 
+// PackageChart creates a chart archive from a path to a chart on the host os and builds chart dependencies
+func (h *Helm) PackageChart(destination string) error {
+	if len(h.Chart.URL) > 0 {
+		url, refPlain, err := transform.GitTransformURLSplitRef(h.Chart.URL)
```
Comment (Noxsios) [inline src/internal/packager/helm/repo.go:30]: Generally go-convention is to not have the package name in any public function names (meh)

File: src/pkg/transform/git.go [hunk 4]
```diff
@@ -12,8 +12,8 @@ import (
 	"github.com/defenseunicorns/zarf/src/pkg/utils/helpers"
 )
 
-// For further explanation: https://regex101.com/r/YxpfhC/3
-var gitURLRegex = regexp.MustCompile(`^(?P<proto>[a-z]+:\/\/)(?P<hostPath>.+?)\/(?P<repo>[\w\-\.]+?)(?P<git>\.git)?(?P<atRef>@(?P<force>\+)?(?P<ref>[\/\+\w\-\.]+))?(?P<gitPath>\/(?P<gitPathId>info\/.*|git-upload-pack|git-receive-pack))?$`)
+// For further explanation: https://regex101.com/r/YxpfhC/5
+var gitURLRegex = regexp.MustCompile(`^(?P<proto>[a-z]+:\/\/)(?P<hostPath>.+?)\/(?P<repo>[\w\-\.]+?)?(?P<git>\.git)?(\/)?(?P<atRef>@(?P<force>\+)?(?P<ref>[\/\+\w\-\.]+))?(?P<gitPath>\/(?P<gitPathId>info\/.*|git-upload-pack|git-receive-pack))?$`)
```
Comment (Noxsios) [inline src/pkg/transform/git.go:16]: Do we have this tested somewhere in a unit test?

Comment (cmwylie19) [reply src/pkg/transform/git.go:16]: more test-cases added. LMK if this is what you had in mind. Thanks!

File: src/internal/packager/helm/repo.go (hunk 3 above)
Comment (cmwylie19) [reply src/internal/packager/helm/repo.go:30]: renamed

File: src/pkg/transform/git.go (hunk 4 above)
Comment (cmwylie19) [reply src/pkg/transform/git.go:16]: Did you want a plan regex case or more [here](https://github.com/defenseunicorns/zarf/blob/0d0cac8496e397ef63adcc276956b4fcdd50811f/src/pkg/transform/git_test.go#L74)?

File: src/pkg/transform/git_test.go [hunk 5]
```diff
@@ -87,6 +88,7 @@ func TestGitTransformURLSplitRef(t *testing.T) {
 		{"https://github.com/defenseunicorns/zarf.helm.git", ""},
 		{"https://github.com/defenseunicorns/zarf.git", "refs/tags/v0.16.0"},
 		{"https://github.com/DoD-Platform-One/big-bang.git", "refs/heads/release-1.54.x"},
+		{"https://github.com/prometheus-community/helm-charts.git", "kube-prometheus-stack-47.3.0"},
```
Comment (Racer159) [inline src/pkg/transform/git_test.go:91]: The two test cases we would need to cover were:

Trailing slash after a path
```
https://github.com/prometheus-community/
```

Trailing slash with no path
```
https://github.com/
```

File: src/pkg/transform/git.go [hunk 6]
```diff
@@ -29,8 +29,8 @@ func MutateGitURLsInText(logger Log, targetBaseURL string, text string, pushUser
 	return output
 }
 
-// GitTransformURLSplitRef takes a git url and returns a separated source url and zarf reference.
-func GitTransformURLSplitRef(sourceURL string) (string, string, error) {
+// GitURLSplitRef takes a git url and returns a separated source url and zarf reference.
+func GitURLSplitRef(sourceURL string) (string, string, error) {
```
Comment (Racer159) [inline src/pkg/transform/git.go:33]: If we do this rename we should do it on all `GitTransform*` functions.

**Review by Racer159 — CHANGES_REQUESTED** [review]

**Review by Noxsios — DISMISSED** [review]

File: src/pkg/transform/git_test.go (hunk 5 above)
Comment (cmwylie19) [reply src/pkg/transform/git_test.go:93]: misinterpreted this somehow, now should be addressed. lmk

**Review by Racer159 — APPROVED** [review]
lgtm

**Pull request merged**